### ISSUE TITLE
What is the best way to play microtonal music?

### ISSUE THREAD

**Issue by ehsansh** [paulrosen/abcjs#636]
Hi Paul,
Thanks for your great work.
We can show quarter sharp and flat by using _/ and ^/ in the music sheet but they are not played as expected.
What is the best way to play for example E a quarter flat? 
Is there any function to change pitch (50 cents lower or upper) of the notes with quarter symbols in whole music sheet?
Or we should make a special soundfont for playing this kind of music?

**Comment by paulrosen** [comment]
It's not supported, but it looks like it would be possible. I'm not sure if I'll have time to play with this, but it looks like when I construct the note the playbackRate needs to be adjusted.

You can modify the soundfont if you don't need the true flat or sharp. For instance, if you want Eb to always be a quarter flat and you don't need D#.

I'll do a quick test and see if the pitch can be modified on the fly.

**Comment by ehsansh** [comment]
Thanks Paul.
I thought to your solution but the problem is that I need E, E flat and also E a quarter flat in one music piece and also if I only need E a quarter flat it seems not logical to make one soundfont for each music piece. 
All different type of notes (flat, sharp, a quarter flat and sharp) are needed in microtonal music. 

I could make a soundfont with all type of notes but the key for each sample is integer could a key be for instance 63.5 to represent a note quarter flat? I don't think so.
There is another possibility I could allocate notes with quarter symbols to keys that are not used for microtonal musical instrument in my soundfont. 
Then I produce two lines of music. For one line I show the animation for other line I play the audio. In audio line I change the position of the notes that have quarter symbols in order to match the soundfont key.

**Comment by paulrosen** [comment]
I think I can make it work without you needing any changes. Maybe soon.

**Comment by paulrosen** [comment]
Try version 6.0.0-beta.31 and let me know if the microtones sound ok. I now am playing a note that is adjusted by 0.02930223664349. I don't know if that is correct or if there is even a standard so let me know if it sounds right on your music.

**Comment by ehsansh** [comment]
Wow! That is great. Thanks a lot.
I could feel Persian music now! I checked it with tuner and it seems correct. 
However in this kind of music quarter flat is not exactly 50 cents lower but now there is huge possibility to use abcjs for folk music. 
Is it possible to add Sori, Koron, Quarter sharp,etc. to the key signature? Like this:
![key-signature](https://user-images.githubusercontent.com/3332025/112049685-94dade80-8b6d-11eb-813a-d9f734ce14d8.png)
It could be great if we could determine which notes of the key signature are a quarter flat, a quarter sharp in the abc string.

**Comment by paulrosen** [comment]
Have you seen this example: http://localhost:63342/abcjs/examples/change_glyphs.html ?
For the key signature, you can specify it exactly, like:
```
K: C ^/f _/B
```

I was thinking that the microtones might not be exactly 50 cents but I have no idea. Do you have a technical reference to the pitches? I also wonder about other traditions. It might be that there needs to be a parameter to tune the microtones.

Also, I'd love to see and hear a real tune in ABC that you've encoded. If you'd like I can use that in change_glyphs and add an audio control to that.

**Comment by ehsansh** [comment]
Thanks Paul! Great job! 
I had seen change_glyphs.html but I didn't know that we are not restricted to have only major, minor scales and we could have any key signature that we like. This is awesome!
There are a lot of books and researches about microtones of Persian music (music of Iran). I know that there are western ethnomusicologists who have published their researches about this kind of music. But I have read them in Persian language and I can't give you a reference. 

About the microtones you are right. As far as I have studied and learned they are not always exactly 50 cents. Different musicians and masters tune their musical instruments differently and when they play a microtone they are different. 
In different [Dastgah ](https://en.wikipedia.org/wiki/Dastgah) the microtones are not the same.  
Even the flat notes like E flat, B flat are not equal in different Dastgah. Sometimes E flat is 10 cents lower than what we have on piano. 

A quarter flat is just a name. The music note could be played 50, 55, 45 cents. I guess in other folk music they have different microtones too. 

What you have made is excellent. If you can add a parameter to tune microtones or maybe other notes in the beginning of the abc string that could be great. Because we can't predict what is the exact microtone that a musician prefer. 

I will send a short tune in Persian music in abc string soon.

**Comment by ehsansh** [comment]
I found that when I click on one music note this function:
```js
ABCJS.synth.playEvent
```
Does not play microtones correctly. Would you please fix it?

And something else that could be helpful if you can change it.
For example for A flat the pitch number in midiPitches array is 68. Could this pitch number for A a quarter flat be 68.5?

**Comment by paulrosen** [comment]
What happens if you have `^/E` and `_/F`? I guess those are the same pitch, the way that `^C` and `_D` are? But if they aren't exactly 50 cents then the might not be the same pitch.

I think there is some complication with using 68.5. The numbers refer to midi pitches. I have created another variable that is the amount of detuning those pitches have. (I realize now that I should have named that variable "cents". I'll change that.) I need to think about what ramifications there are for not having the pitch be a midi number.

**Comment by ehsansh** [comment]
I have not seen `^/E` or ` ^/B` yet. 
Maybe the reason is that in Persian music the smallest interval that we have in 99.99% of the cases is a semitone. 
We tune for instance A a quarter flat. But when we play the first tetrachord of Shur which is the most important Dastgah in Persian music the notes are like this: `G  _/A _B c`  
If we play Chahargah the first tetrachord of this Dastgah is `G _/A B c`
As you can see the interval between the notes of one tetrachord could be half tone, 3 quarter tone, 5 quarter tone , etc.
if you like to hear how it sounds with its original musical instrument let's have a look on [this web app](https://rameshgar.com/santoor/pieces/467/%D8%A8%D8%B1%D9%81-%D9%BE%DB%8C%D8%B1%DB%8C-%D8%AA%D9%88-%D8%A8%D8%AE%D9%88%D8%A7%D9%86-%D8%A2%D9%85%D9%88%D8%B2%D8%B4-%D8%B3%D9%86%D8%AA%D9%88%D8%B1) that I have made. Please click the yellow play button in the bottom of the page.
I am trying to learn abc to make such a tune with abcjs.

**Comment by ehsansh** [comment]
Maybe you have not seen my first request.
`_/A` is played correctly when I click the play button and all the tune is playing. But when I click on these notes on the music sheet `_/A` and `A` have the same pitch. Would you please have a look on this?

**Comment by ehsansh** [comment]
Of course I am satisfied with your great work now. Thank you.
I think maybe detuning and adding pitch variables is not a high priority.

**Comment by paulrosen** [comment]
I added a demo and I needed a representative tune. After a quick search I randomly found this:
```
X:1
T:مجنون نبودم
%%graceslurs 0
L:1/4
M:2/4
K: Bb ^/A
F/G/ G/G/ | Gd|c/d/ B/c/|cd|c/d/ B/c/|
Bc|B/c/ A/B/|Gz|F,/G,/ G,/G,/|G,D| 
C/D/ B,/C/| CD|C/D/ B,/C/|B,C|B,/C/ A,/B,/|
G,z/D/|CB,/C/|B, z/D/|CB,/C/|B,zD/|
CB,/C/|{B,}(A,{G,})z/C/|B,A,/B,/|AG|:F/G//G// G/G/|
Gd|c/c//d// B/c/|cd|c/c//d// B/c/|Bc|
B/B//c// A/B/ |1Gz:|2G z/d/||cB/c/|Bz/d/|
cB/c/|Bz/d/|cB/c/|{B}(A {G})z/c/|BA/B/|
AG/C/|Gz|]
```
I don't know if that is correct transcription or a traditional tune. If you have one that would be better for the example I'd love to get it. If that is an appropriate tune, what does the title mean and is there a history to the tune?

**Comment by ehsansh** [comment]
Oh, sorry Paul I was so late to send a tune. I am not fast enough in writing abc string. 
This piece, Majnun nabudam ( مجنون نبودم ), is good. I checked and it is acceptable but the key signature should be changed to `K: Bb _/A` and it is better to eliminate those grace notes. As it has lyric it is better to have a lower tempo. 
The title means: I was not insane.
The poet in this tune says to his beloved: I was not insane. You made me insane. You drove me out of town. Which mountain has your scent? 

I am not a poem translator! 

I am not aware of its history. In folk music of Iran we have many tunes but even we don't know who has composed them. Maybe this is something that in centuries has been composed, modified many times.
Of course there are many tunes composed in the last century by great composers they could be great examples too.

**Comment by paulrosen** [comment]
I just wanted to make sure it was traditional and not problematic. Fixing the key signature made a huge difference - it now makes musical sense to me.

**Comment by paulrosen** [comment]
Fixed in 6.0.0-beta.33
